torchtune/models/llama3_2_vision/_convert_weights.py

```diff
@@ -171,3 +237,188 @@ def llama3_vision_tune_to_meta(
             value = value.reshape(dim, -1)
         converted_state_dict[new_key] = value
     return converted_state_dict
+
+
+def llama3_vision_hf_to_tune(
+    state_dict: Dict[str, torch.Tensor],
+    num_heads: int = 32,
+    num_kv_heads: int = 32,
+    dim: int = 4096,
+    head_dim: int = None,
+    vocab_size: int = 128256,
+    cross_attention_layers: Optional[List[int]] = None,
+    # Vision Encoder Paramters
+    encoder_dim: int = 1280,
+    tile_size: int = 448,
+    num_tiles: int = 4,
+    supported_aspect_ratios: List[Tuple[int, int]] = None,
+) -> Dict[str, torch.Tensor]:
+    """
+    Convertor from HF state dict to torchtune state dict. This handles:
+    - Updateing the cross attention layer numbers
+    - skip loading the rope embeddings
+    - reshaping q, k projections
+    - reversing the precomputed vision positional embeddings
+    """
+    converted_state_dict = {}
+    if head_dim is None:
+        head_dim = dim // num_heads
+    if cross_attention_layers is None:
+        cross_attention_layers = []
+
+    def _permute(t, n_heads):
+        return (
+            t.view(n_heads, 2, head_dim // 2, dim)
+            .transpose(1, 2)
+            .reshape((head_dim * n_heads), dim)
+        )
+
+    for key, value in state_dict.items():
+        if "rotary_emb.inv_freq" in key:  # Skip loading the position embeddings
+            continue
+        new_key = get_mapped_key(key, _FROM_HF)
+        if "language_model" in key:
+            if "layers" in key:  # Update layer numbers
+                layer = int(key.split(".")[3])
```

Comment:
Not the end of the world, but I feel like half the reason we added the placeholders in e.g. `_FROM_HF` was so that we could avoid stuff like this (e.g. [here](https://github.com/pytorch/torchtune/blob/10b02e0e358952205ab1f6eb7fcbf24b99d4b7fd/torchtune/models/convert_weights.py#L48-L65)) and keep the `hf_to_tune` methods simpler. I know the multimodal models are more involved, so maybe there's no way around it. But still, all of this is a bit hard to parse

Reply:
There are a number of challenges here that our setup doesn't handle out of the box and I'm not quite sure if there's a good generalization. The layer numbering scheme for cross_attention is different between ours and theirs. There are also a number of instances where a single one of their keys maps to two of our keys. We could solve the latter generally, but I'm not sure about the former, we might need some layer mapping too. If we add more MM models, we might want to go down that route.